internal/extproc/translator/openai_gcpanthropic.go

```diff
@@ -6,59 +6,715 @@
+			return sb.String()
+		}
+	default:
+		return ""
+	}
+	return ""
+}
+
+func anthropicRoleToOpenAIRole(role anthropic.MessageParamRole) (string, error) {
+	switch role {
+	case anthropic.MessageParamRoleAssistant:
+		return openai.ChatMessageRoleAssistant, nil
+	case anthropic.MessageParamRoleUser:
+		return openai.ChatMessageRoleUser, nil
+	default:
+		return "", fmt.Errorf("invalid anthropic role %v", role)
+	}
+}
 
-// RequestBody implements [Translator.RequestBody] for GCP Anthropic.
-// This method translates an OpenAI ChatCompletion request to a GCP Anthropic API request.
-func (o *openAIToGCPAnthropicTranslatorV1ChatCompletion) RequestBody(_ []byte, openAIReq *openai.ChatCompletionRequest, onRetry bool) (
+// openAIMessageToAnthropicMessageRoleAssistant converts an OpenAI assistant message to Anthropic content blocks.
+// The tool_use content is appended to the Anthropic message content list if tool_calls are present.
+func openAIMessageToAnthropicMessageRoleAssistant(openAiMessage *openai.ChatCompletionAssistantMessageParam) (anthropicMsg *anthropic.MessageParam, err error) {
+	contentBlocks := make([]anthropic.ContentBlockParamUnion, 0)
+	if v, ok := openAiMessage.Content.Value.(string); ok && len(v) > 0 {
+		contentBlocks = append(contentBlocks, anthropic.NewTextBlock(v))
+	} else if content, ok := openAiMessage.Content.Value.(openai.ChatCompletionAssistantMessageParamContent); ok {
+		switch content.Type {
+		case openai.ChatCompletionAssistantMessageParamContentTypeRefusal:
+			if content.Refusal != nil {
+				contentBlocks = append(contentBlocks, anthropic.NewTextBlock(*content.Refusal))
+			}
+		case openai.ChatCompletionAssistantMessageParamContentTypeText:
+			if content.Text != nil {
+				contentBlocks = append(contentBlocks, anthropic.NewTextBlock(*content.Text))
+			}
+		default:
+			err = fmt.Errorf("content type not supported: %v", content.Type)
+			return
+		}
+	}
+
+	// Handle tool_calls (if any).
+	for i := range openAiMessage.ToolCalls {
+		toolCall := &openAiMessage.ToolCalls[i]
+		var input map[string]interface{}
+		if err = json.Unmarshal([]byte(toolCall.Function.Arguments), &input); err != nil {
+			err = fmt.Errorf("failed to unmarshal tool call arguments: %w", err)
+			return
+		}
+		toolUse := anthropic.ToolUseBlockParam{
+			ID:    toolCall.ID,
+			Type:  "tool_use",
+			Name:  toolCall.Function.Name,
+			Input: input,
+		}
+		contentBlocks = append(contentBlocks, anthropic.ContentBlockParamUnion{OfToolUse: &toolUse})
+	}
+
+	return &anthropic.MessageParam{
+		Role:    anthropic.MessageParamRoleAssistant,
+		Content: contentBlocks,
+	}, nil
+}
+
+// openAIToAnthropicMessages converts OpenAI messages to Anthropic message params type, handling all roles and system/developer logic.
+func openAIToAnthropicMessages(openAIMsgs []openai.ChatCompletionMessageParamUnion) (anthropicMessages []anthropic.MessageParam, systemBlocks []anthropic.TextBlockParam, err error) {
+	for i := range openAIMsgs {
+		msg := openAIMsgs[i]
```

Comment:
This will be (really) slightly efficient since it won't copy the value

```suggestion
		msg := &openAIMsgs[i]
```